lib/rouge/lexers/sas.rb

```diff
@@ -0,0 +1,568 @@
+        rule %r(/[*].*?[*]/)m, Comment::Multiline
+
+        # date/time constants (Language Reference pp91-2)
+        rule /'[0-9a-z]+?'d/i, Literal::Date
+        rule /'.+?'dt/i, Literal::Date
+        rule /'[0-9:]+?([a|p]m)?'t/i, Literal::Date
+
+        rule /'/, Str::Single, :single_string
+        rule /"/, Str::Double, :double_string
+        rule /&[a-z0-9_&.]+/i, Name::Variable
+
+        # numeric constants (Language Reference p91)
+        rule /\d[0-9a-f]*x/i, Num::Hex
+        rule /\d[0-9e\-.]+/i, Num # scientific notation
+
+        # auto variables from DATA step (Language Reference p46, p37)
+        rule /\b(_n_|_error_|_file_|_infile_|_msg_|_iorc_|_cmd_)\b/i, Name::Builtin::Pseudo
+
+        # auto variable list names
+        rule /\b(_character_|_numeric_|_all_)\b/i, Name::Builtin
+
+        # datalines/cards etc
+        rule /\b(datalines|cards)(\s*)(;)/i do
+          groups Keyword, Text, Punctuation
+          push :datalines
+        end
+        rule /\b(datalines4|cards4)(\s*)(;)/i do
+          groups Keyword, Text, Punctuation
+          push :datalines4
+        end
+
+
+        # operators (Language Reference p96)
+        rule %r(\*\*|[\*/\+-]), Operator
+        rule /[^¬~]?=:?|[<>]=?:?/, Operator
+        rule /\b(eq|ne|gt|lt|ge|le|in)\b/i, Operator::Word
+        rule /[&|!¦¬∘~]/, Operator
+        rule /\b(and|or|not)\b/i, Operator::Word
+        rule /[<>|><]/, Operator # min/max
+        rule /\|\|/, Operator # concatenation
+
+        # The OF operator should also be highlighted (Language Reference p49)
+        rule /\b(of)\b/i, Operator::Word
+        rule /\b(like)\b/i, Operator::Word # Language Ref p181
+      
+        rule /\d+/, Num::Integer
+
+        rule /\$/, Keyword::Type
+
+        # Macro definitions
+        rule /(%macro|%mend)(\s*)([\w\d]+)/i do
+          groups Keyword, Text, Name::Function
+        end
+        rule /%mend/, Keyword
+
+        rule /%\w[\w\d]*/ do |m|
+          if self.class.sas_macro_statements.include? m[0].upcase
+            token Keyword
+          elsif self.class.sas_macro_functions.include? m[0].upcase
+            token Keyword
+          else
+            token Name
+          end
+        end
+      end
+
+      state :basics2 do
+        # Rules to be parsed after the keywords (which are different depending
+        # on the context)
+
+        # Missing values (Language Reference p81)
+        rule /\s\.[;\s]/, Keyword::Constant # missing
+        rule /\s\.[a-z_]/, Name::Constant # user-defined missing
+
+        rule /[\(\),;:\{\}\[\]\\\.]/, Punctuation
+
+        rule /@/, Str::Symbol # line hold specifiers
+        rule /\?/, Str::Symbol # used for format modifiers
+        
+        rule /.*/, Text # Fallback for anything we haven't matched so far
+      end
+
+
+
+      state :root do
+        mixin :basics
+
+        # PROC definitions
+        rule %r((proc)(\s+)(\w+))ix do |m|
+          @proc_name = m[0].split(' ')[1].downcase
```

Comment:
This should be a local variable rather than an instance variable, right?

Reply:
I was going for an instance variable so that when in `state :proc` we would know which PROC we were in and could retrieve the relevant keywords for that. Happy to amend if there's a better way of doing it!

Reply:
Ah, I see what you mean. My initial reaction is that this kind of checking isn't really the type of thing to do in the lexer, you could just groups all the keywords together and if that means some false positives, it shouldn't be the end of the world.

Now that you've gone and done all the hard work, though, I'm happy to leave as-is.

Reply:
Oh, one thing, though is that you shouldn't need to split `m[0]` since you're capturing the terms separately. You should be able to get the keyword from `m[3]`.

Reply:
Thanks; fixed in https://github.com/rouge-ruby/rouge/pull/1107/commits/f0a0871123d7062290d9734d26a0fd66c6fd4fc1 . For simplicity I also changed the keys of the hash to upper case -- I was previously doing `downcase` then `upcase` again!